server/vulnerabilities/nvd/cve.go

```diff
@@ -272,3 +285,157 @@ func checkCVEs(
+	// CPEMatch rules can include version strings for the following:
+	// - versionStartIncluding
+	// - versionStartExcluding
+	// - versionEndExcluding
+	// - versionEndIncluding - not used in this function as we don't want to assume the resolved version
+	for _, rule := range cpeMatch {
+		if rule.VersionEndExcluding == "" {
+			continue
+		}
+
+		// convert the NVD cpe23URi to wfn.Attributes for later comparison
+		attr, err := wfn.Parse(rule.Cpe23Uri)
+		if err != nil {
+			return "", ctxerr.Wrap(ctx, err, "parsing cpe23Uri")
+		}
+
+		// ensure the product and vendor match
+		if attr.Product != hostSoftwareMeta.Product || attr.Vendor != hostSoftwareMeta.Vendor {
+			continue
+		}
+
+		// versionEnd is the version string that the vulnerable host software version must be less than
+		versionEnd, err := checkVersion(ctx, rule, softwareVersion)
+		if err != nil {
+			return "", ctxerr.Wrap(ctx, err, "checking version")
+		}
+		if versionEnd != "" {
+			return versionEnd, nil
+		}
+	}
+
+	return "", nil
+}
+
+// CPEMatch can be nested in Children nodes. Recursively search the nodes for a CPEMatch
+func findCPEMatch(nodes []*schema.NVDCVEFeedJSON10DefNode) []*schema.NVDCVEFeedJSON10DefCPEMatch {
+	for _, node := range nodes {
+		if len(node.CPEMatch) > 0 {
+			return node.CPEMatch
+		}
+
+		if len(node.Children) > 0 {
+			match := findCPEMatch(node.Children)
+			if match != nil {
+				return match
+			}
+		}
+	}
+	return nil
+}
+
+// checkVersion checks if the host software version matches the CPEMatch rule
+func checkVersion(ctx context.Context, rule *schema.NVDCVEFeedJSON10DefCPEMatch, softwareVersion *semver.Version) (string, error) {
+	for _, condition := range []struct {
+		startIncluding string
+		startExcluding string
+	}{
+		{rule.VersionStartIncluding, ""},
+		{"", rule.VersionStartExcluding},
+	} {
+		constraintStr := buildConstraintString(condition.startIncluding, condition.startExcluding, rule.VersionEndExcluding)
+		if constraintStr == "" {
+			return rule.VersionEndExcluding, nil
+		}
+
+		constraint, err := semver.NewConstraint(constraintStr)
+		if err != nil {
+			return "", ctxerr.Wrapf(ctx, err, "parsing constraint: %s", constraintStr)
+		}
+
+		if constraint.Check(softwareVersion) {
+			return rule.VersionEndExcluding, nil
+		}
+	}
+
+	return "", nil
+}
+
+// buildConstraintString builds a semver constraint string from the startIncluding,
+// startExcluding, and endExcluding strings
+func buildConstraintString(startIncluding, startExcluding, endExcluding string) string {
+	if startIncluding == "" && startExcluding == "" {
+		return ""
+	}
+	if startIncluding != "" {
+		return fmt.Sprintf(">= %s, < %s", startIncluding, endExcluding)
+	}
+	return fmt.Sprintf("> %s, < %s", startExcluding, endExcluding)
+}
+
+// Products using 4 part versioning scheme (ie. docker desktop)
+// need to be converted to 3 part versioning scheme (2.3.0.2 -> 2.3.0+3) for use with
+// the semver library.
+func preprocessVersion(version string) string {
```

Comment:
Could we have unit tests for this method?

Reply:
good idea, these vendor versions are getting out of hand, updated